dglgo/dglgo/pipeline/graphpred/graphpred.jinja-py

```diff
@@ -0,0 +1,133 @@
+
+def train(device, loader, model, criterion, optimizer):
+    model.train()
+
+    for _, (g, labels) in enumerate(tqdm(loader, desc="Iteration")):
+        g = g.to(device)
+        labels = labels.reshape(g.batch_size, -1).to(device)
+        node_feat = g.ndata['feat']
+        edge_feat = g.edata['feat']
+
+        pred = model(g, node_feat, edge_feat)
+        optimizer.zero_grad()
+        # ignore nan targets (unlabeled) when computing training loss
+        is_labeled = labels == labels
+        loss = criterion(pred.float()[is_labeled], labels.float()[is_labeled])
+        loss.backward()
+        optimizer.step()
+
+def calc_metric(y_true, y_pred):
+    task_metric_list = []
+    for i in range(y_true.shape[1]):
+        # AUC is only defined when there is at least one positive and negative datapoint.
+        if np.sum(y_true[:, i] == 1) > 0 and np.sum(y_true[:, i] == 0) > 0:
+            # ignore nan values
+            is_labeled = y_true[:,i] == y_true[:,i]
+            task_metric = sklearn.metrics.{{ user_cfg.general_pipeline.metric }}(
+                y_true[is_labeled, i], y_pred[is_labeled, i])
+            task_metric_list.append(task_metric)
+
+    return sum(task_metric_list) / len(task_metric_list)
+
+def evaluate(device, loader, model):
+    model.eval()
+    y_true = []
+    y_pred = []
+
+    for _, (g, labels) in enumerate(tqdm(loader, desc="Iteration")):
+        g = g.to(device)
+        labels = labels.reshape(g.batch_size, -1).to(device)
+        node_feat = g.ndata['feat']
+        edge_feat = g.edata['feat']
+
+        with torch.no_grad():
+            pred = model(g, node_feat, edge_feat)
+        y_true.append(labels.view(pred.shape).detach().cpu())
+        y_pred.append(pred.detach().cpu())
+
+    y_true = torch.cat(y_true, dim=0).numpy()
+    y_pred = torch.cat(y_pred, dim=0).numpy()
+
+    return calc_metric(y_true, y_pred)
+
+def main(run):
+    {{ user_cfg_str }}
+    device = cfg['device']
+    if not torch.cuda.is_available():
+        device = 'cpu'
+    pipeline_cfg = cfg['general_pipeline']
+    save_path = pipeline_cfg['save_path']
+
+    # load data
+    data = AsGraphPredDataset({{data_initialize_code}})
+    train_loader = GraphDataLoader(data[data.train_idx], batch_size=pipeline_cfg['train_batch_size'],
+                                   shuffle=True, num_workers=pipeline_cfg['num_workers'])
+    val_loader = GraphDataLoader(data[data.val_idx], batch_size=pipeline_cfg['eval_batch_size'],
+                                 shuffle=False, num_workers=pipeline_cfg['num_workers'])
+    test_loader = GraphDataLoader(data[data.test_idx], batch_size=pipeline_cfg['eval_batch_size'],
+                                  shuffle=False, num_workers=pipeline_cfg['num_workers'])
+
+    # create model
+    model_cfg = cfg["model"]
+    # data[0] is a tuple (g, label)
+    cfg["model"]["data_info"] = {
+        "name": cfg["data_name"],
+        "node_feat_size": data[0][0].ndata['feat'].shape[-1],
```

Comment:
Consider add `node_feat_size` and `edge_feat_size` as class attributes of `AsGraphPredDataset`.